libsupport: initial logging and result classes

ddn0 opened the pull request:
Ok. I've updated this PR according to the feedback:

# Major changes
- Long names: GALOIS_LOG_XYZ
- Levels: Error, Warning, Verbose, Debug. Verbose is only printed when GALOIS_LOG_VERBOSE
    is set in the environment. Debug is only printed if -DCMAKE_BUILD_TYPE=Debug. This is a slight change from the semantics suggested by @witchel where Verbose was conditionally compiled in and Debug was controlled by runtime configuration, but I think the current naming is more clear (Debug is enabled in Debug builds).
- Result type class has been removed from this PR
- galois::EnvCheck (from libgalois) has been copied here as galois::GetEnv (libsupport) in order to support GALOIS_LOG_VERBOSE ; after this lands, all uses of galois::EnvCheck should just use galois::GetEnv instead.
- People like fmt so I added fmt support; this will pose some fun when this code gets upstreamed. (@insertinterestingnamehere I'm thinking of installing fmt as a submodule upstream since y'all don't use conan?)

# Merge Plan
Since everyone who is interested in this library is here, let's finish the review here, and once this is approved, I'll port it over to iss/Galois.

<del>
Design Proposal via PR: happy to take comments.

This commit adds logging functions that do not depend on the Galois threading library, which should make it more suitable for wider use.

Also this commit adds a `galois::Result` type that can be used by functions to explicitly return errors. This is an alternative to throwing exceptions and I think it should be the preferred way to indicate errors going forward. It forces callers to deal with errors. It is explicit about which functions may fail. And it is straightforward model for developers to manage and extend.

The `galois::Result` and `galois::Status` classes are copy-pasted from Arrow, which in turn was copy-pasted from LevelDB/Kudo/... We are in an illustrious chain of engineering. Compared to the original Arrow versions of these files, I removed a lot of `Status`es and unlikely branch annotations from the original code. The latter was introduced to alleviate some performance issues: https://issues.apache.org/jira/browse/ARROW-2400 Given the evolving state of compilers, I'd prefer to start simple and performance tune for our own use-cases.

There is a reference implementation of the yet-to-be-standardized `std::expected` proposal which is an alternative to the `Result` class:  https://github.com/TartanLlama/expected but there still seems to be some small warts like  https://github.com/TartanLlama/expected/issues/78 and it doesn't seem widely used in the libraries we use so we wouldn't get the benefit of consistent cross-API error handling. I opted to do the safest thing here and manage this ourselves.

@insertinterestingnamehere : I don't know if you think this should be upstreamed as well. If so, libsupport might be a good place to move macros, etc. 
</del>

tylershunt commented on libsupport/include/galois/Logging.h:
Might be nice to add auto `__FILE__` and `__LINE__` to debug messages as well (or maybe all Log messages if compiled in Debug mode?). I guess this is a personal taste thing; not a deal breaker for me.

tylershunt commented on libsupport/include/galois/Logging.h:
woops didn't realize this was still a draft; forgive an overzealous reviewer.

ddn0 commented on libsupport/include/galois/Logging.h:
I wouldn't mind that either; though there is some tension between using the preprocessor (which gives us `__FILE__` and `__LINE__`) and avoiding preprocessor edge cases like `GALOIS_FATAL(..., templated_function<type1, type2>(...))`. The comma `,` will be processed by the preprocessor instead of the C++ compiler.

In the Arrow code, they did something like this:
```cpp
ARROW_LOG(LEVEL) << "message";
```
where `#define ARROW_LOG(level) writer(__FILE__, __LINE__)` which made it a little easier to separate the processor bit from the standard C++ world. That doesn't seem too bad to me. Though I'm not a hige fan of the `<<` operator or C++ stream API. 

If people have better clean ways to getting `__FILE__` and `__LINE__` in messages, I'd love to do it.

ddn0 commented on libsupport/include/galois/Logging.h:
@tylershunt I certainly want comments! I'd take your approval as a :+1:

tylershunt commented on libsupport/include/galois/Logging.h:
Not sure I understand the corner case, but in the past I've done things like:
`#define GALOIS_LOG(level, ...) galois::Log ## level (__FILE__ ":" #__LINE__, __VA_ARGS__)`
(didn't compile so not 100% on the syntax)

rossbach commented on libsupport/include/galois/Logging.h:
I'm a fan of anything that uses and deals cleanly with __FILE__ and __LINE__. Like Donald, I'm not a fan of the C++ stream API, so the ARROW_LOG variant on all this doesn't get me super-excited. Both Tyler's suggestion and the approach illustrated by GALOIS_FATAL(...) work for me.

ddn0 commented on libsupport/include/galois/Logging.h:
> It seems like this doesn't solve your corner case.

Now that I think about it, my issue only arises if the original macro had arguments:
```cpp
#define MACRO(one, two) ...
MACRO(template_function<1, 2>(), "my_two")
```
here one = `template_function<1` and two = `2>()` and "my_two" gives us a preprocessor error because the macro invocation doesn't match the definition. It doesn't seem to arise with `__VA_ARGS__` so I think taking the macro path as the main entrypoint (to get `__FILE__` and co everywhere) seems ok here.

tylershunt commented on libsupport/include/galois/Logging.h:
💯 That approach has my vote.

insertinterestingnamehere commented on libsupport/include/galois/Macros.h:
Could this just be `[[nodiscard]]`?

ddn0 commented:
Some details of `boost::outcome` https://www.boost.org/doc/libs/1_70_0/libs/outcome/doc/html/faq.html

tylershunt commented:
> Most of our compiler-specific configuration info usually ends up in headers, so the library structure doesn't make a huge difference one way or another.

I think it does matter for the katana (if not for upstream) since ideally we'd like to call some of those macros in `tsuba`. Linking a common support library seems better than special casing some headers (or header directories).

witchel commented:
> Some details of `boost::outcome` https://www.boost.org/doc/libs/1_70_0/libs/outcome/doc/html/faq.html (not boost version: https://ned14.github.io/outcome/faq/)
> 
> FWIW, I'm leaning towards `boost::outcome` over the current `galois::Result` class. Though, reducing our boost dependencies to at most the header-only libraries is a personal goal of mine :D

That FAQ sure was detailed.  Like, solely detailed.

Are boost::outcome and logging levels dijoint issues?  Because it would be nice to specify the log level of different messages.

insertinterestingnamehere commented:
> I think it does matter for katana (if not for upstream) since ideally we'd like to call some of those macros in `tsuba`. Linking a common support library seems better than special casing some headers (or header directories).

Great, upstreaming this is fine with me if it's more convenient for others.

ddn0 commented:
> Are boost::outcome and logging levels dijoint issues? Because it would be nice to specify the log level of different messages.

Totally disjoint. I joined them into one PR because I was lazy (and these two types of features are good candidates for a support library).

My current thinking is:
- getting rid of `galois::Result` in favor of `boost::outcome`
- a series of macros, `GALOIS_LOG_DEBUG`, `GALOIS_LOG_INFO`, `GALOIS_LOG_ERROR`, `GALOIS_FATAL` all of which will report `__FILE__` and `__LINE__` information.

An alternative to separate macros would be `GALOIS_LOG(level, ...)` and `GALOIS_FATAL`. Is that what you were suggesting @witchel ? I find that using more than a few levels hard to keep consistent in a codebase, which is why I'm suggesting a separate macro for each level. Though I can go either way.

I do like structured logging (where it is easy to add ad-hoc fields to messages), but that doesn't seem too common in C++ codebases.

insertinterestingnamehere commented:
WRT the long-term goal of getting rid of the boost dependency, we could use https://ned14.github.io/outcome/ instead. I think it's the same library as `boost::outcome`, just not packaged inside boost. Boost is probably easier for now though.

witchel commented:
> > Are boost::outcome and logging levels dijoint issues? Because it would be nice to specify the log level of different messages.
> 
> Totally disjoint. I joined them into one PR because I was lazy (and these two types of features are good candidates for a support library).
> 
> My current thinking is:
> 
> * getting rid of `galois::Result` in favor of `boost::outcome`
> * a series of macros, `GALOIS_LOG_DEBUG`, `GALOIS_LOG_INFO`, `GALOIS_LOG_ERROR`, `GALOIS_FATAL` all of which will report `__FILE__` and `__LINE__` information.
> 
> An alternative to separate macros would be `GALOIS_LOG(level, ...)` and `GALOIS_FATAL`. Is that what you were suggesting @witchel ? I find that using more than a few levels hard to keep consistent in a codebase, which is why I'm suggesting a separate macro for each level. Though I can go either way.
> 
> I do like structured logging (where it is easy to add ad-hoc fields to messages), but that doesn't seem too common in C++ codebases.

I don't think I need arbitrary levels, but I like the diversity and terseness of Android logging function names: Log.v(), Log.d(), Log.i(), Log.w(), and Log.e() for verbose, debugging, info, warn and error.  Here is the explanation: The order in terms of verbosity, from least to most is ERROR, WARN, INFO, DEBUG, VERBOSE. Verbose should never be compiled into an application except during development. Debug logs are compiled in but stripped at runtime. Error, warning, and info logs are always kept.

I could live without INFO.

Should this macro be in the GALOIS namespace?

ddn0 commented:
> I don't think I need arbitrary levels, but I like the diversity and terseness of Android logging function names

I'm pretty ambivalent about the names; would people like to vote their preference here?
:+1: short names, e.g., `GALOIS_LOG_V`
:-1: long names, .e.g, `GALOIS_LOG_VERBOSE`

ddn0 commented:
> Verbose should never be compiled into an application except during development. Debug logs are compiled in but stripped at runtime. Error, warning, and info logs are always kept.

This seems pretty reasonable; I can get behind these semantics.

ddn0 commented:
> Should this macro be in the GALOIS namespace?

The alternative would be a new `KATANA` namespace or having a namespace per lib-level "project"?

I think while https://github.com/intelligentsoftwaresystems/galois is our upstream, we should aim to be compatible with them. I think post the Galois 6.0 release and as we need to add features that aren't really topics of research, we should consider stronger forking. Though `s/GALOIS/KATANA/` just for branding seems like a lot of noise for not too much clarity.

Lib-level, e.g., `SUPPORT`, seems too fine-grained for me.

roshandathathri commented:
> * a series of macros, `GALOIS_LOG_DEBUG`, `GALOIS_LOG_INFO`, `GALOIS_LOG_ERROR`, `GALOIS_FATAL` all of which will report `__FILE__` and `__LINE__` information

How is this different from galois::runtime::trace(), galois::gDebug(), galois::gInfo(), galois::gWarn(), galois::gError()? Is the only difference that you're adding file and line information?

witchel commented:
> > Should this macro be in the GALOIS namespace?
> 
> The alternative would be a new `KATANA` namespace or having a namespace per lib-level "project"?

Yes, you are convincing.

ddn0 commented:
> How is this different from galois::runtime::trace(), galois::gDebug(), galois::gInfo(), galois::gWarn(), galois::gError()? Is the only difference that you're adding file and line information?

These functions require libgalois_shmem so they cannot be used by libraries that libgalois_shmem itself may depend on, like libtsuba, and any other lower-level libraries we might want to write.

roshandathathri commented:
> These functions require libgalois_shmem so they cannot be used by libraries that libgalois_shmem itself may depend on, like libtsuba, and any other lower-level libraries we might want to write.

I got that. My question was about functionality/usage (apart from the library dependence). 

In addition, why not pull those functions out of libgalois_shmem instead of re-implementing it?

ddn0 commented:
> My question was about functionality/usage (apart from the library dependence).

So far, there is some consensus towards consistent line printing, which isn't being done by the `galois::gDebug` family of functions.

> In addition, why not pull those functions out of libgalois_shmem instead of re-implementing it?

It is hard to say what it means to just pull these functions from libgalois_shmem because as they are written they cannot be separated from libgalois_shmem. They depend on `getTID()`, etc.

I personally would prefer that DEBUG or its equivalent not write to a file and that it should not query current time on every print or if it does all the logging functions should. Though, no write to file and no timestamp is my preference for a general log function.

roshandathathri commented:
> I personally would prefer that DEBUG or its equivalent not write to a file and that it should not query current time on every print or if it does all the logging functions should. Though, no write to file and no timestamp is my preference for a general C++ log function.

I agree with what you said and the discussion above. I would prefer if the functionality is not duplicated and we just have one function to do one thing. There are two three cases: general/single-thread logging, shared-memory logging, and distributed-memory logging, and each can have distinct names like galois::log, galois::parallelLog, and galois::distLog respectively. Each can have the 5 different variants of logging. All of them can use the same sub-routines and shared-memory can add thread ID while distributed-memory can add host ID. I'm okay with dropping shared-memory logging. What do you think?

ddn0 commented:
> There are two three cases: general/single-thread logging, shared-memory logging, and distributed-memory logging, and each can have distinct names like galois::log, galois::parallelLog, and galois::distLog respectively. Each can have the 5 different variants of logging. All of them can use the same sub-routines and shared-memory can add thread ID while distributed-memory can add host ID. I'm okay with dropping shared-memory logging. What do you think?

I think distributed logging is sufficiently different that it could live outside libsupport. For example, timestamps, host ids, logging to file make a lot more sense for distributed programs than single host ones.

I don't think it's worth another set of functions just for the shared-memory case. If people want thread IDs they can add them themselves. (Frankly, I consider using `getTID` or showing the thread ID an advance use case already.) We are already going to make logging thread-safe so the libsupport version should suffice for both general purpose and shared-memory.

If that's agreeable, we should plan to land libsupport with logging and with the semantics @witchel suggested. The name vote is not clearly decisive (though I have been taken with the shorter names), so I propose sticking with the longer names because that matches the style of the codebase in general and the previous logging functions.

Then, we can clean up shared-memory usage to use the new functions.

Separately from all the above, we can talk about dist logging. It doesn't seem like that is on the critical path, and we could leave things as-is. Though, correct me if I'm wrong. From my inexperienced eye, it seems messy with both Tracer.cpp and gIO.cpp but maybe things are good enough.

I can see how having a more structured and uniform output would let one visualize communication more easily. I've seen timelines like https://horovod.readthedocs.io/en/stable/summary_include.html#horovod-timeline used with good effect (all this requires is emitting a particular type of JSON document and loading it in your browser).

ddn0 commented on libsupport/include/galois/Macros.h:
Yes it can. `boost::outcome` actually uses that exact annotation.

witchel commented:
> Though I'm not a huge fan of the << operator or C++ stream API.

Have you considered fmt?
https://github.com/fmtlib/fmt

roshandathathri commented:
> I think distributed logging is sufficiently different that it could live outside libsupport.

Yes. libsupport can provide the common sub-routines and general logging. Both shared-memory logging (if needed) and distributed-memory logging can be built on top of libsupport. 

> it seems messy with both Tracer.cpp and gIO.cpp

Yes, I'm in favor of dropping tracer. If we have a verbose level of log, that would be a good replacement for tracer.

> I've seen timelines like https://horovod.readthedocs.io/en/stable/summary_include.html#horovod-timeline used with good effect

Thanks for sharing this. This seems like a nice feature to have for distributed logging.

insertinterestingnamehere commented:
> > Though I'm not a huge fan of the << operator or C++ stream API.
> 
> Have you considered fmt?
> https://github.com/fmtlib/fmt

Coming to the standard library in C++20 https://en.cppreference.com/w/cpp/utility/format.

ddn0 commented:
>> Have you considered fmt?
> Coming to the standard library in C++20 https://en.cppreference.com/w/cpp/utility/format.

fmt is great for general string writing.

I don't know if it is worth the extra dependency for logging in particular though. If we are going through the effort to make messages easy to read for humans, I'd also like those messages to be easy to be read by machines as well, which is why I prefer some support for structured logging over better support for general strings.

tylershunt commented:
> which is why I prefer some support for structured logging over better support for general strings.

👍 for structured logging.

witchel commented:
> fmt is great for general string writing.
> 
> I don't know if it is worth the extra dependency for logging in particular though. If we are going through the effort to make messages easy to read for humans, I'd also like those messages easy to be read by machines as well, which is why I prefer some support for structured logging over better support for general strings.

Yes, yes.  I was just writing output from my benchmark program and bemoaning the lack of python's f-strings in c++ when I saw that they were coming to c++20, but not yet present in our environment.  

I like structured logging and while my deployment experience is vintage 90s, what about allowing logging format control from environment variables?

ddn0 commented:
> I like structured logging and while my deployment experience is vintage 90s, what about allowing logging format control from environment variables?

I suppose the question is along the gradient from (a) `printf` with `__FILE__` and `__LINE__` at one end to (b) https://www.boost.org/doc/libs/1_73_0/libs/log/doc/html/index.html where do we want to put libsupport logging.

I was leaning closer to (a) because in this repo we have libraries (functionality designed to be integrated into another library or application) and applications. Libraries benefit from some lightweight debugging printf support "this works" and simple assertions like "unexpected state, die and print a line number". Anything more complicated probably should be reflected in a return value or error code rather than a log message IMHO.

Applications might need a lot more since their output is user/operator-facing. If someone was building a server in C++, I would totally suggest something like `boost::log`.

witchel commented:
> I was leaning closer to (a)

I'm picking up what you are laying down.  The middle road sounds good.

The storage layer can always wrap these functions with our own (where we can use our overly terse names).

You'll have to pardon my enthusiasm.  Most of my fantasy life these days involve creative but obscure choices for environment variables that enable in-situ debugging of sticky customer problems.

ddn0 commented:
FWIW, here is an application-level logging package that does `fmt` strings, logging levels from the environment, ring-buffer and configurable output sinks: https://github.com/gabime/spdlog

(AFAIK, no structured fields though)

tylershunt commented:
> (AFAIK, no structured fields though)

Yeah I haven't found anything super promising on that front either; I'd be happy with these fmt strings; or even just `*f` versions of the logging macros that let me do c style formating.

ddn0 commented:
Ok. I've updated this PR according to the feedback:

Major changes:
- Long names: `GALOIS_LOG_XYZ`
- Levels: Error, Warning, Verbose, Debug. Verbose is only printed when `GALOIS_LOG_VERBOSE`
is set in the environment. Debug is only printed if `-DCMAKE_BUILD_TYPE=Debug`. This is a slight change from the semantics suggested by @witchel where Verbose was conditionally compiled in and Debug was controlled by runtime configuration, but I think the current naming is more clear (Debug is enabled in Debug builds).
- Result type class has been removed from this PR
- `galois::EnvCheck` (from libgalois) has been copied here as `galois::GetEnv` in order to support `GALOIS_LOG_VERBOSE` ; after this lands, all uses of `galois::EnvCheck` should just use `galois::GetEnv` instead.
- People like `fmt` so I added `fmt` support; this will pose some fun when this code gets upstreamed.

witchel commented:
> Thanks for wrangling!

Yes, thank you.  Logging is super vulnerable to bike-shedding.

roshandathathri commented on libsupport/include/galois/Logging.h:
What is the purpose of do-while in all these cases? Is it to handle the semi-colon after GALOIS_*()?

roshandathathri commented on libsupport/src/Logging.cpp:
Why not templatize the first two arguments?

roshandathathri commented on libsupport/test/logging.cpp:
I'm assuming this C++-style printing like the current gDebug()?

ddn0 commented on libsupport/include/galois/Logging.h:
It is actually to handle this case:

```cpp
if (blah)
  GALOIS_LOG_ERROR(...)
```

(without braces)

ddn0 commented on libsupport/test/logging.cpp:
This is a C++20-style `fmt` specification: https://en.cppreference.com/w/cpp/utility/format/format backported via https://github.com/fmtlib/fmt

I don't believe `gDebug()` supports this style of printing.

For example:
```
GALOIS_LOG_ERROR("basic format string {}", 42);
```
would print
```
basic format string
```

I'm playing around with the compile-time checking for the format strings. It does support type-checking arguments, e.g., `fmt::format(FMT_STRING("basic format string {:s}"), 42)` is a compile-time error while `fmt::format(FMT_STRING("basic format string {:d}"), 42)` would succeed, but it doesn't raise an error if something is missing like in ``fmt::format(FMT_STRING("basic format string"), 42)`

roshandathathri commented on libsupport/include/galois/Logging.h:
That can be handled with just braces (without do-while) right?

ddn0 commented on libsupport/src/Logging.cpp:
If people are fine with all logging going to stderr, we can just remove the first two arguments since they are invariant in this file.

In general, I'd prefer avoiding templating things just because we can (code bloat, readability, maintainability). In this particular case, we are printing things out so optimizing away a branch or two doesn't seem super important at this stage (unless you have faced a concrete bottleneck logging?). This code is also under a lock.

roshandathathri commented on libsupport/src/Logging.cpp:
Ok, that makes sense.

insertinterestingnamehere commented:
> * People like `fmt` so I added `fmt` support; this will pose some fun when this code gets upstreamed. (@insertinterestingnamehere I'm thinking of installing `fmt` as a submodule upstream since y'all don't use conan?)

Most distros already have packages for fmt at this point though so we can probably pull it in the same way we do boost and llvm. We use lmod modules for dependency management internally and can build a copy from source for that. We should be able to add this dependency after the 6.0 release.

Honestly, I'm not a fan of using submodules for dependencies, but if people want that, it's okay as long as there's a way to override the vendored version of the library with an installed one. Libraries with vendored dependencies that involve user-facing APIs or ABIs and that don't have overrides to pull in installed versions from elsewhere are an awful pain to package.

insertinterestingnamehere commented on libsupport/src/Logging.cpp:
It'd be nice to be able to configure where the output goes. On Windows, powershell by default treats writing to stderr as an indicator that a command has failed, so the best practice there is to only write to stderr if you're planning to exit the program with a nonzero exit status. I know we don't support that environment right now, but I'm still hoping we'll get that working someday.

insertinterestingnamehere commented on libsupport/src/Logging.cpp:
> It'd be nice to be able to configure where the output goes. On Windows, powershell by default treats writing to stderr as an indicator that a command has failed, so the best practice there is to only write to stderr if you're planning to exit the program with a nonzero exit status. I know we don't support that environment right now, but I'm still hoping we'll get that working someday.

Note: as far as I'm aware, the general consensus is that this was a dumb decision on the part of the powershell developers, but there's no indicator it'll ever change so I usually try to avoid running into trouble with it.

ddn0 commented on libsupport/src/Logging.cpp:
> It'd be nice to be able to configure where the output goes. 

At runtime or compile time? It seems like how Windows/Powershell treats the console is different than how a Unix-y system would. It would be too weird to me to just print to standard out on all windows builds.

ddn0 commented on libsupport/include/galois/Logging.h:
https://stackoverflow.com/questions/257418/do-while-0-what-is-it-good-for

ddn0 commented:
> Honestly, I'm not a fan of using submodules for dependencies, but if people want that, it's okay as long as there's a way to override the vendored version of the library with an installed one.

Nor I.

Ok. I'll plan for ` libfmt-dev` to be installed externally then (in the iss/Galois repo).

insertinterestingnamehere commented on libsupport/src/Logging.cpp:
Yah, this particular way of doing things is way different from other OS's and shells. The right thing to do on Windows though is to only write to stderr if you're writing an error message. Logs of any other kind should go somewhere else (probably just stdout). It doesn't really matter *when* we make that distinction during the compilation process. It just matters that where the output goes is configurable.

insertinterestingnamehere commented on libsupport/src/Logging.cpp:
Actually, the powershell stderr weirdness is arguably a reason to not make this kind of thing user-configurable because then the logging library can just take care of things itself. We could do something like: on Windows fatal and error go to stderr and everything else goes to stdout, on other systems everything goes to stderr.

ddn0 commented:
Ok. Last major change.

I've added compile time checking of `fmt` strings. It doesn't cover the case of missing arguments, but it does some type-checking, e.g., https://github.com/KatanaGraph/katana/pull/75#discussion_r440961619

This does mean that all `fmt` strings must be `constexpr`s or string constants. That doesn't seem that bad of a limitation in exchange for some type-safety for `fmt` arguments (and it probably more readable), but please let me know if that is too onerous a restriction.

tylershunt commented:
Bump; I think this is settled right?

arthurp reviewed (approved):
It appears the only semantic change is `~FileView` becoming virtual. Is that true?

Otherwise LGTM.

ddn0 commented:
> It appears the only semantic change is ~FileView becoming virtual. Is that true?

Yup.

Off the top of my head, I don't know if mixing `override` and `virtual` is a linter problem or a Werror issue.

The `const` stuff is just because having that as a separate PR seems too tedious, but I can split it out if you want :D

arthurp commented:
Carry on. I just wanted to validate that most of this is just updating to "override" instead of "virtual".
